cumulus/pallets/xcmp-queue/src/benchmarking.rs

```diff
@@ -17,12 +17,156 @@
 
 use crate::*;
 
-use frame_benchmarking::{benchmarks, impl_benchmark_test_suite};
+use codec::DecodeAll;
+use frame_benchmarking::v2::*;
+use frame_support::traits::Hooks;
 use frame_system::RawOrigin;
 
-benchmarks! {
-	set_config_with_u32 {}: update_resume_threshold(RawOrigin::Root, 100)
-	set_config_with_weight {}: update_weight_restrict_decay(RawOrigin::Root, Weight::from_parts(3_000_000, 0))
-}
+#[benchmarks]
+mod benchmarks {
+	use super::*;
+
+	/// Modify any of the `QueueConfig` fields with a new `u32` value.
+	///
+	/// Used as weight for:
+	/// - update_suspend_threshold
+	/// - update_drop_threshold
+	/// - update_resume_threshold
+	#[benchmark]
+	fn set_config_with_u32() {
+		#[extrinsic_call]
+		Pallet::<T>::update_resume_threshold(RawOrigin::Root, 1);
+	}
+
+	#[benchmark]
+	fn enqueue_xcmp_message() {
+		assert!(QueueConfig::<T>::get().drop_threshold * MaxXcmpMessageLenOf::<T>::get() > 1000);
+		let msg = BoundedVec::<u8, MaxXcmpMessageLenOf<T>>::default();
+
+		#[block]
+		{
+			Pallet::<T>::enqueue_xcmp_message(0.into(), msg, &mut WeightMeter::new()).unwrap();
+		}
+	}
+
+	#[benchmark]
+	fn suspend_channel() {
+		let para = 123.into();
+		let data = ChannelSignal::Suspend.encode();
+
+		#[block]
+		{
+			ChannelSignal::decode_all(&mut &data[..]).unwrap();
+			Pallet::<T>::suspend_channel(para);
+		}
+
+		assert_eq!(
+			OutboundXcmpStatus::<T>::get()
+				.iter()
+				.find(|p| p.recipient == para)
+				.unwrap()
+				.state,
+			OutboundState::Suspended
+		);
+	}
+
+	#[benchmark]
+	fn resume_channel() {
+		let para = 123.into();
+		let data = ChannelSignal::Resume.encode();
+
+		Pallet::<T>::suspend_channel(para);
+
+		#[block]
+		{
+			ChannelSignal::decode_all(&mut &data[..]).unwrap();
+			Pallet::<T>::resume_channel(para);
+		}
+
+		assert!(
+			OutboundXcmpStatus::<T>::get().iter().find(|p| p.recipient == para).is_none(),
+			"No messages in the channel; therefore removed."
+		);
+	}
+
+	/// Split a singular XCM.
+	#[benchmark]
+	fn split_concatenated_xcm() {
+		let max_downward_message_size = MaxXcmpMessageLenOf::<T>::get() as usize;
+
+		// Assumes `xcm::MAX_INSTRUCTIONS_TO_DECODE`: 100
+		let max_instrs = 100 - MAX_XCM_DECODE_DEPTH;
+		let mut xcm = Xcm::<T>(vec![ClearOrigin; max_instrs as usize]);
 
-impl_benchmark_test_suite!(Pallet, crate::mock::new_test_ext(), crate::mock::Test);
+		for _ in 0..MAX_XCM_DECODE_DEPTH - 1 {
+			xcm = Xcm::<T>(vec![Instruction::SetAppendix(xcm)]);
```

Comment:
Does this work now that we sum up all instructions to get to the max? Including nested ones